pennylane/tracker.py

```diff
@@ -0,0 +1,166 @@
+
+    >>> tracker.totals
+    {'executions': 3, 'shots': 300}
+    >>> tracker.history
+    {'executions': [1, 1, 1], 'shots': [100, 100, 100]}
+    >>> tracker.latest
+    {'executions': 1, 'shots': 100}
+
+    .. UsageDetails::
+
+    .. note::
+        With backpropagation, this function should take ``qnode.device``
+        instead of the device used to create the QNode.
+
+    Users can pass a custom callback function to the ``callback`` keyword. This
+    function is run each time the ``record()`` method is called. The
+    function passed must accept ``totals``, ``history``, and ``latest`` as
+    keyword arguments:
+
+    >>> def shots_info(totals=dict(), history=dict(), latest=dict()):
+    ...     print("Total shots: ", totals['shots'])
+    >>> with qml.Tracker(circuit.device, callback=shots_info) as tracker:
+    ...     qml.grad(circuit)(0.1)
+    Total shots:  100
+    Total shots:  200
+    Total shots:  300
+
+    By specifying ``persistent=False``, you can reuse the same tracker accross
+    multiple contexts.
+
+    >>> with qml.Tracker(circuit.device, persistent=False) as tracker:
+    ...     circuit(0.1)
+    >>> with tracker:
+    ...     circuit(0.2)
+    >>> tracker.totals['executions']
+    2
+
+    """
+
+    def __init__(self, dev=None, callback=None, persistent=False):
+        self.persistent = persistent
+
+        self.callback = callback
+
+        # same code as self.reset
+        self.totals = dict()
+        self.history = dict()
+        self.latest = dict()
+
+        self.tracking = False
+
+        if dev is not None:
+            if not dev.capabilities().get("supports_tracker", False):
+                raise Exception(f"Device '{dev.short_name}' does not support device tracking")
+            dev.tracker = self
+
+    def __enter__(self):
+        if not self.persistent:
+            self.reset()
+
+        self.tracking = True
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        self.tracking = False
+
+    def update(self, **kwargs):
+        """Store passed keyword-value pairs into ``totals``,``history``, and ``latest`` attributes.
+
+        There is no restriction on the key-value pairs passed.
+
+        >>> tracker.update(a=1, b=2)
+        >>> tracker.latest
+        {"a":1, "b":2}
+        >>> tracker.history
+        {"a": [1], "b": [2]}
+        >>> tracker.totals
+        {"a": 1, "b": 2}
+
+        """
+        self.latest = kwargs
+
+        for key, value in kwargs.items():
+            # update history
+            if key in self.history.keys():
+                self.history[key].append(value)
+            else:
+                self.history[key] = [value]
+
+            # updating totals
+            if value is not None:
+                # Only total numeric values
+                try:
+                    self.totals[key] = value + self.totals.get(key, 0)
+                except TypeError:
+                    pass
```

Comment:
This line doesn't seem to be covered.

Reply:
I'm testing the functionality and it behaves the way I want it to.

Basically, `totals` is incremented if the value can be summed with `0`.  If there's a better way of doing this, I would happily change the try-except block, but I can't come up with anything better.

Reply:
How about `isinstance(self.totals.get(key, 0), numbers.Number)` or a more extended check if `numbers.Number` doesn't cover all possibilities (e.g., not sure about numpy scalars)?

If it doesn't come around, then a test case would be if there's a key in `self.totals` which is not numeric, such that a `TypeError` is raised.

Reply:
What if the value is `np.array(0)`? That's not a number, but can be summed with zero.

Maybe I just start with the more restrictive `isinstance(value, numbers.Number)`.  If we ever start running into a use-case where we need more generality, we can change the code.

Reply:
@albi3ro that is a really good point...

It perhaps convinces me further that your `try-except` solution is superior 🙂